File: tools/gen_struct.py
import os
import sys


def snake_to_camel(snake_str):
    parts = snake_str.split('_')
    return ''.join(word.title() for word in parts)


head_mapping = {
	'参数名': 'name',
	'字段': 'name',
	'类型': 'type',
	'内容': 'content',
	'必要性': 'notnull',
	'备注': 'comment',
}


if __name__ == '__main__':
	while True:
		head = []
		lines = []
		# 从控制台读取输入
		print('请输入Markdown表格，在最后一行之后输入ok表示结束：（退出请输入exit）')
		while True:
			line = input()
			if not line or line.strip() == 'exit':
				sys.exit(0)
			if line.strip() == 'ok':
				break
			if len(head) == 0:
				head = line.split('|')
				for i in range(len(head)):
					v = head[i].strip()
					head[i] = head_mapping.get(v, '')
			else:
				l = line.split('|')
				if len(l) != len(head):
					print('列数不一致，无法解析')
					sys.exit(0)
				m = {}
				for i in range(len(l)):
					v = l[i].strip()
					if v.startswith('---'):
						m = None
						break
					m[head[i]] = v
				if not m:
					continue
				lines.append(m)
		# 生成结构体
		print('type T struct {')
		for m in lines:
			name = snake_to_camel(m['name'])
			if m['type'] == 'num':
				m['type'] = 'int'
			elif m['type'] == 'str':
				m['type'] = 'string'
			elif m['type'] == 'bool':
				pass
			elif m['type'] == 'array':
				m['type'] = '[]' + name
			elif m['type'] == 'obj':
				m['type'] = name
			if m.get('notnull', '必要') == '必要':
				m['notnull'] = ''
			else:
				m['notnull'] = ',omitempty'
			content = m.get('content', '')
			comment = m.get('comment', '')
			if content or comment:
				sep = '。' if content and comment else ''
				comment = ' // %s%s%s' % (content, sep, comment)
				comment = comment.replace('<br>', '。').replace('<br/>', '。').replace('<br />', '。')
			print('\t%s %s `json:"%s%s"`%s' % (name, m['type'], m['name'], m['notnull'], comment))
		print('}')

File: tools/gen_struct.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func snakeToCamel(snakeStr string) string {
	parts := strings.Split(snakeStr, "_")
	for i, word := range parts {
		parts[i] = strings.Title(word) // nolint: staticcheck
	}
	return strings.Join(parts, "")
}

var headMapping = map[string]string{
	"参数名": "name",
	"字段":  "name",
	"字段名": "name",
	"项":   "name",
	"类型":  "type",
	"内容":  "content",
	"必要性": "notnull",
	"备注":  "comment",
}

func main() {
	for {
		var head []string
		var lines []map[string]string
		fmt.Println("请输入Markdown表格，在最后一行之后输入ok表示结束：（退出请输入exit）")
		scanner := bufio.NewScanner(os.Stdin)
		for scanner.Scan() {
			line := scanner.Text()
			if line == "" {
				continue
			}
			if line == "exit" {
				os.Exit(0)
			}
			if line == "ok" {
				break
			}
			line = strings.ReplaceAll(line, "`", "")
			if len(head) == 0 {
				head = strings.Split(line, "|")
				for i, v := range head {
					head[i] = strings.TrimSpace(v)
					head[i] = headMapping[head[i]]
				}
			} else {
				l := strings.Split(line, "|")
				if len(l) != len(head) {
					fmt.Println("列数不一致，无法解析")
					os.Exit(0)
				}
				m := make(map[string]string)
				for i, v := range l {
					v = strings.TrimSpace(v)
					if strings.HasPrefix(v, "---") {
						m = nil
						break
					}
					m[head[i]] = v
				}
				if m != nil {
					lines = append(lines, m)
				}
			}
		}
		fmt.Println("type T struct {")
		for _, m := range lines {
			name := snakeToCamel(m["name"])
			switch m["type"] {
			case "num":
				m["type"] = "int"
			case "str":
				m["type"] = "string"
			case "bool":
			case "array", "Array", "list", "List", "array(obj)", "Array(obj)", "list(obj)", "List(obj)":
				m["type"] = "[]"
				if !strings.HasSuffix(name, "s") {
					m["type"] += name
				} else {
					m["type"] += name[:len(name)-1]
				}
			case "array(num)", "Array(num)", "list(num)", "List(num)":
				m["type"] = "[]int"
			case "array(str)", "Array(str)", "list(str)", "List(str)":
				m["type"] = "[]string"
			case "array(bool)", "Array(bool)", "list(bool)", "List(bool)":
				m["type"] = "[]bool"
			case "obj":
				m["type"] = name
			}
			notNullValue := `"`
			if m["notnull"] != "必要" && m["notnull"] != "必须" && m["notnull"] != "必填" && m["notnull"] != "√" {
				notNullValue = `,omitempty" request:"query,omitempty"`
			}
			content := m["content"]
			comment := m["comment"]
			sep := ""
			if content != "" && comment != "" {
				sep = "。"
			}
			commentText := fmt.Sprintf("// %s%s%s", content, sep, comment)
			commentText = strings.ReplaceAll(commentText, "<br>", "。")
			commentText = strings.ReplaceAll(commentText, "<br/>", "。")
			commentText = strings.ReplaceAll(commentText, "<br />", "。")
			commentText = strings.ReplaceAll(commentText, `\`, "")
			fmt.Printf("\t%s %s `json:\"%s%s`%s\n", name, m["type"], m["name"], notNullValue, commentText)
		}
		fmt.Println("}")
	}
}
